cloverage/src/cloverage/source.clj

```diff
@@ -30,34 +49,24 @@
         resource-path (-> (.getResource (get-loader) resource) .toURI .getPath)]
     (str/replace resource-path cwd "")))
 
-(defn resource-reader [resource]
+(defn resource-reader ^java.io.Closeable [resource]
   (if-let [resource (and resource
                          (.getResourceAsStream
                           (get-loader)
                           resource))]
     (InputStreamReader. resource) ; We assume the default charset is set correctly
     (throw (IllegalArgumentException. (str "Cannot find resource " resource)))))
 
-(defn form-reader [ns-symbol]
+(defn form-reader ^java.io.Closeable [ns-symbol]
   (if-let [res-path (resource-path ns-symbol)]
     (rt/indexing-push-back-reader
      (PushbackReader.
       (resource-reader res-path)))
     (throw (ex-info (format "Resource path not found for namespace: %s" (name ns-symbol))
                     {:ns-symbol ns-symbol}))))
 
-(defn forms [ns-symbol]
-  (let [src (form-reader ns-symbol)]
-    (loop [forms nil]
-      (if-let [form (r/read src false nil)]
-        (recur (conj forms form))
-        (reverse forms)))))
-
-(defn ns-form [ns-symbol]
-  (let [src (form-reader ns-symbol)]
-    (first (drop-while #(not= 'ns (first %))
-                       (take-while (comp not nil?)
-                                   (repeatedly #(r/read {:eof nil
-                                                         :features #{:clj}
-                                                         :read-cond :allow}
-                                                        src)))))))
+(defn forms
+  "Return all forms in the source file for the namespace named by `ns-symbol`."
+  [ns-symbol]
+  (with-open [reader (form-reader ns-symbol)]
+    (doall (take-while (partial not= ::eof) (repeatedly #(r/read reader false ::eof))))))
```

Comment:
the old implementation of `forms` was leaky -- it wasn't using `with-open` -- so I replaced it with a more concise implementation. It does the same thing